// automation-events/build/node/module.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  AutomationEventList: true,
  createCancelAndHoldAutomationEvent: true,
  createCancelScheduledValuesAutomationEvent: true,
  createExponentialRampToValueAutomationEvent: true,
  createLinearRampToValueAutomationEvent: true,
  createSetTargetAutomationEvent: true,
  createSetValueAutomationEvent: true,
  createSetValueCurveAutomationEvent: true
};
Object.defineProperty(exports, "AutomationEventList", {
  enumerable: true,
  get: function () {
    return _automationEventList.AutomationEventList;
  }
});
Object.defineProperty(exports, "createCancelAndHoldAutomationEvent", {
  enumerable: true,
  get: function () {
    return _createCancelAndHoldAutomationEvent.createCancelAndHoldAutomationEvent;
  }
});
Object.defineProperty(exports, "createCancelScheduledValuesAutomationEvent", {
  enumerable: true,
  get: function () {
    return _createCancelScheduledValuesAutomationEvent.createCancelScheduledValuesAutomationEvent;
  }
});
Object.defineProperty(exports, "createExponentialRampToValueAutomationEvent", {
  enumerable: true,
  get: function () {
    return _createExponentialRampToValueAutomationEvent.createExponentialRampToValueAutomationEvent;
  }
});
Object.defineProperty(exports, "createLinearRampToValueAutomationEvent", {
  enumerable: true,
  get: function () {
    return _createLinearRampToValueAutomationEvent.createLinearRampToValueAutomationEvent;
  }
});
Object.defineProperty(exports, "createSetTargetAutomationEvent", {
  enumerable: true,
  get: function () {
    return _createSetTargetAutomationEvent.createSetTargetAutomationEvent;
  }
});
Object.defineProperty(exports, "createSetValueAutomationEvent", {
  enumerable: true,
  get: function () {
    return _createSetValueAutomationEvent.createSetValueAutomationEvent;
  }
});
Object.defineProperty(exports, "createSetValueCurveAutomationEvent", {
  enumerable: true,
  get: function () {
    return _createSetValueCurveAutomationEvent.createSetValueCurveAutomationEvent;
  }
});

var _automationEventList = require("./classes/automation-event-list");

var _createCancelAndHoldAutomationEvent = require("./functions/create-cancel-and-hold-automation-event");

var _createCancelScheduledValuesAutomationEvent = require("./functions/create-cancel-scheduled-values-automation-event");

var _createExponentialRampToValueAutomationEvent = require("./functions/create-exponential-ramp-to-value-automation-event");

var _createLinearRampToValueAutomationEvent = require("./functions/create-linear-ramp-to-value-automation-event");

var _createSetTargetAutomationEvent = require("./functions/create-set-target-automation-event");

var _createSetValueAutomationEvent = require("./functions/create-set-value-automation-event");

var _createSetValueCurveAutomationEvent = require("./functions/create-set-value-curve-automation-event");

var _index = require("./interfaces/index");

Object.keys(_index).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _index[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _index[key];
    }
  });
});

var _index2 = require("./types/index");

Object.keys(_index2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _index2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _index2[key];
    }
  });
});
